mart/attack/model_transform.py

```diff
@@ -0,0 +1,38 @@
+#
+# Copyright (C) 2022 Intel Corporation
+#
+# SPDX-License-Identifier: BSD-3-Clause
+#
+
+from ..utils import MonkeyPatch
+
+
+class LightningModuleAsTarget:
+    """Prepare a LightningModule as a target model for Adversary,
+    such that `output = model(batch)`.
+    """
+
+    def __call__(self, model):
+        # Generate a pseudo dataloader_idx.
+        dataloader_idx = 1
+
+        if hasattr(model, "attack_step"):
+
+            def model_forward(batch):
+                output = model.attack_step(batch, dataloader_idx)
+                return output
+
+        elif hasattr(model, "training_step"):
+            # Monkey-patch model.log to avoid spamming.
+            def model_forward(batch):
+                with MonkeyPatch(model, "log", lambda *args, **kwargs: None):
+                    output = model.training_step(batch, dataloader_idx)
+                return output
+
+        else:
+            raise ValueError("Model does not have `attack_step()` or `training_step()`.")
+
+        return model_forward
```

Comment:
This logic should be able to live in `Adversary`.

Reply:
Separating the logic here should simplify `Adversary` if we are going to keep `Adversary.model_transform` anyway. 
`model_transform` should be handy if we want to change a model before attack, e.g. freezing the BN statistics, or changing the non-differentiable operations.

Reply:
Thanks for the suggestion. I moved the logic to `Adversary` because that makes it less verbose. We need to define `unittest.mock.Mock` carefully then, because `hasattr(Mock(), "any_string") == True`. 1d7606147e3321aaf0665ede4b8562c6d53ad43d